### PR TITLE
Upgrade outdated packages; Bump React to v18

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary
Previously we attempted to upgrade React in #7394 but due to React version inconsistencies on the upstream, we ended up blocking that PR. 

This PR aims to upgrade React to v18 with the following enhancements:
- Migrated JS(React) test cases to `@testing-library/react`.
- Update the `AfterActivationSiteScan` service to work with React >= 18.
-  Migrated from deprecated `registerStore` in the `@wordpress/data` package.
- Add jest resolver to fix failing tests due to the UUID package. 
- Updated tests container in wp-env commands.
- Update workflow to run E2E tests with Node 16.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: src/Admin/AfterActivationSiteScan.php [hunk 1]
```diff
@@ -84,13 +92,17 @@ public function __construct( RESTPreloader $rest_preloader ) {
...
+		$react = wp_scripts()->query( self::REACT );
+
 		return (
 			is_admin()
 			&&
 			Services::get( 'dependency_support' )->has_support()
 			&&
 			! is_network_admin()
 			&&
+			( $react instanceof _WP_Dependency && version_compare( $react->ver, '18', '>=' ) )
```
Comment (thelovekesh) [inline src/Admin/AfterActivationSiteScan.php:104]: @westonruter Please note the `AfterActivationSiteScan` service will only work with React 18 and will not show the Site Scan Notice on plugin/themes activation if WP < 6.2.

Comment (westonruter) [reply src/Admin/AfterActivationSiteScan.php:104]: Unless the latest version of the Gutenberg plugin is active, right?

Comment (thelovekesh) [reply src/Admin/AfterActivationSiteScan.php:104]: Yes. The latest Gutenberg plugin uses React 18, so if it's activated then it will work on WP < 6.2.

File: assets/src/block-validation/components/error/test/error.js [hunk 2]
```diff
@@ -243,11 +229,18 @@ describe('Error', () => {
 				)
 			).not.toBeNull();
 
-			container
-				.querySelector(
+			// container
+			// 	.querySelector(
+			// 		`.amp-error--${getErrorTypeClassName(status)} button`
+			// 	)
+			// 	.click();
```
Comment (westonruter) [inline assets/src/block-validation/components/error/test/error.js]: Why is this commented-out code here?

File: assets/src/block-validation/components/with-amp-toolbar-button/test/add-toolbar-button-to-block.js [hunk 3]
```diff
@@ -38,39 +38,18 @@ describe('withAMPToolbarButton: filtering with errors', () => {
...
-					clientId: block.clientId,
-					code: 'DISALLOWED_TAG',
-					status: 3,
-					term_id: 12,
-					title: 'Invalid script: <code>jquery.js</code>',
-					type: 'js_error',
-				},
-			],
-		});
-	});
```
Comment (westonruter) [inline assets/src/block-validation/components/with-amp-toolbar-button/test/add-toolbar-button-to-block.js:64]: I recall we chatted about this a bit, but I see why the change of `status` to `1`? Why are `unreviewedValidationErrors` and `validationErrors` seemingly merged?

File: assets/src/components/conditional-details/test/index.js [hunk 4]
```diff
@@ -48,26 +27,20 @@ describe('ConditionalDetails', () => {
...
-					{[null, null]}
-				</ConditionalDetails>,
-				container
-			);
-		});
+		({ container } = render(
+			<ConditionalDetails summary={<div>{'Summary'}</div>}>
+				{[null, null]}
+			</ConditionalDetails>
+		));
```
Comment (westonruter) [inline assets/src/components/conditional-details/test/index.js:43]: Why the parentheses wrapping this statement?

File: .github/workflows/build-test-measure.yml [hunk 5]
```diff
@@ -331,6 +331,14 @@ jobs:
         env:
           COMPOSE_INTERACTIVE_NO_CLI: true
 
+
+      # TODO: Remove it once node version in .nvmrc is updated to >=16.
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:335]: Wait, why isn't `.nvmrc` updated to node 18 if the `engines.node` was updated in `package.json`?

**Review by westonruter — COMMENTED** [review]
What an epic lift! Great work. Just a couple small comments, and aside from the failing tests, looks really good. (Granted, I'm not an expert on React!)

File: assets/src/block-validation/components/error/test/error.js (hunk 2 above)
Comment (thelovekesh) [reply assets/src/block-validation/components/error/test/error.js]: This code has been migrated to use the click event from `@testing-library/react` and I missed removing it. Will remove it 👍🏼

File: assets/src/block-validation/components/with-amp-toolbar-button/test/add-toolbar-button-to-block.js (hunk 3 above)
Comment (thelovekesh) [reply assets/src/block-validation/components/with-amp-toolbar-button/test/add-toolbar-button-to-block.js:64]: Previously we were using `registerStore` to create redux stores and then initializing them at required places. With this previously, we were able to initialize it with the initial state of our choice. 

With the current implementation, we are initializing our store itself where it has been defined so there will be a single instance of store everywhere. We have combined both of them in one because it's the job of the reducer to decide which validation error will be assigned to unreviewed or reviewed as per

https://github.com/ampproject/amp-wp/blob/9f7014a73e3c16adc310df1575e6d8d4a9435e44/assets/src/block-validation/store/index.js#L82-L123

Previously, `unreviewedValidationErrors` was being passed manually(which shouldn't be the case) to check if we are getting the expected results. By passing 1 as status the reducer automatically adds the validation error to the `unreviewedValidationErrors` state.

Also, see how Gutenberg creates stores: https://github.com/WordPress/gutenberg/blob/trunk/packages/notices/src/store/index.js

File: .github/workflows/build-test-measure.yml (hunk 5 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:335]: Aah, I missed reverting the changes in `engines`. Since many deps in `@wordpress/*` are outdated, let's keep node to 14.

File: assets/src/components/conditional-details/test/index.js (hunk 4 above)
Comment (thelovekesh) [reply assets/src/components/conditional-details/test/index.js:43]: To allow destructuring without a declaration as per [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#assignment_separate_from_declaration_2:~:text=Note%3A%20The%20parentheses%20(%20...%20)%20around%20the%20assignment%20statement%20are%20required%20when%20using%20object%20literal%20destructuring%20assignment%20without%20a%20declaration.).

<details><summary>Example</summary>

```js
const obj = {
	a: 1,
	b: 2,
};

const obj2 = {
	a: 2,
	b: 3,
};

let { a } = obj;

({ a } = obj2); // It will work without declaring `let a` but ESLint will give an error, so assigning it. 

console.log(a); // 2

```

</details>

**Review by westonruter — APPROVED** [review]
🎉

**Review by westonruter — APPROVED** [review]
🎉

**Comment by thelovekesh** [comment]
QA Passed ✅ 

- All JS Unit tests are passing after being migrated to `@testing-library/react`.
- `AfterActivationSiteScan` service only works when React 18 support is present.
